src/scripts/util.sh

```diff
@@ -37,31 +37,164 @@ error() {
+# wildcards.
 #
-# $1: Path to a Nginx configuration file.
-parse_server_names() {
-    sed -n -r -e 's&^\s*server_name\s+(.*);.*&\1&p' "$1" | xargs -n1 echo | uniq
+# $1: The path (support for wildcards) to match configuration files
+# $2: An associative bash array that will contain cert_name => server_names
+#     (space-separated) after the call to this function
+parse_config_files_for_certs() {
+    local -n certs=$2
+
+    for conf_file in $1; do
+        # To follow if we are in a server block, and how to match the ending of that
+        # server block (we expect equal indentation at the beginning and end)
+        local in_server_block=0
+        local leading_server_block_spaces=
+
+        # The resources we're looking for, that we will reset between server blocks
+        local cert_names=()
+        local server_names=()
+
+        local lineno=0
+        while IFS="" read -r line || [ -n "$line" ]; do
+            lineno=$((lineno + 1))
+            if [[ "$line" =~ ^([[:space:]]*)server[[:space:]]*\{ ]]; then
+                if [ $in_server_block -eq 1 ]; then
+                    error "Matching server block while already in a server block ($conf_file:$lineno)"
+                    # Do we want to go to next file, exit or keep parsing ?
+                fi
+                # We entered a server block, so we'll need to start checking what's up
+                in_server_block=1
+                leader_server_block_spaces=${BASH_REMATCH[1]}
+            elif [ $in_server_block -eq 0 ]; then
+                # We are not in a server block, nothing for us to do here, let's skip
+                continue
+            elif [[ "$line" =~ ^${leading_server_block_spaces}\} ]]; then
+                # We add the found certificate names and the corresponding server names,
+                # in case it already exists, this will be appended nicely as if it was
+                # all in the same server block. In case there was no 'ssl_certificate_key'
+                # entry, we can simply ignore it nicely.
+                if [ ${#server_names[@]} -eq 0 ]; then
+                    error "Missing 'server_name' for server block ($conf_file:$lineno)"
+                elif [ ${#cert_names[@]} -gt 0 ]; then
+                    for cert_name in "${cert_names[@]}"; do
+                        certs["$cert_name"]="${certs["$cert_name"]}${certs["$cert_name"]:+ }${server_names[@]}"
+                    done
+                fi
+
+                # We are leaving a server block, so we can clean-up behind us
+                in_server_block=0
+                leadin_server_block_spaces=
+
+                cert_names=()
+                server_names=()
+            elif [[ "$line" =~ ^[[:space:]]*ssl_certificate_key[[:space:]]*/etc/letsencrypt/live/([^;]*)/privkey.pem[[:space:]]*\; ]]; then
+                cert_names+=(${BASH_REMATCH[1]})
+            elif [[ "$line" =~ ^[[:space:]]*server_name[[:space:]]*([^\;]*)[[:space:]]*\; ]]; then
```

Comment:
I've been hesitating for this: fo we want to make a better pattern, or to filter after ? None is done for now.

I believe we should filter after.
However, how do we want to handle that ? Ignore any `server_name` line that contains something we cannot get a certificate for (like fake wildcards, as nginx supports them but SSL certificates don't), or ignore only that entry and keep the rest of the other `server_names` on that same line ? (like `*.example.com example.com something*` would only keep `*.example.com` and `example.com` for instance)

Reply:
I actually went ahead and implemented the filtering "after the fact".
I've just printed debug messages for now, if we want to raise that to warnings, pretty easy :)